### PR TITLE
[SE-0293] Implement revision #3 of property wrapper parameters

### PR DESCRIPTION
**Pull request by hborla**
* The distinction between API wrappers and implementation-detail wrappers is formalized, and determined by the compiler based on whether the property wrapper type defines `init(projectedValue:)` (and also defines `var projectedValue`).
* Only API property wrappers on parameters use caller-side application of the property wrapper, and are part of the function signature. Implementation-detail property wrappers on parameters use callee-side application of the property wrapper, and have no external effect on the function.

### CODE REVIEW

File: include/swift/AST/PropertyWrappers.h [hunk 1]
```diff
@@ -170,16 +170,36 @@ void simple_display(llvm::raw_ostream &os, PropertyWrapperLValueness l);
...
+      : backingVar(backingVar), projectionVar(projectionVar) {}
+
+  /// Whether this is a valid property wrapper.
+  bool isValid() const {
+    return backingVar != nullptr;
+  }
+
+  explicit operator bool() const { return isValid(); }
+
+  friend bool operator==(const PropertyWrapperAuxiliaryVariables &lhs,
```
Comment (slavapestov) [inline include/swift/AST/PropertyWrappers.h:194]: We no longer use the AnyValue type eraser to store cached request values so unless you need it for some other purpose, you don’t have to overload operator== anymore.

Comment (hborla) [reply include/swift/AST/PropertyWrappers.h]: I removed it, thanks!

**Comment by hborla** [comment]
@swift-ci please build toolchain

**Comment by hborla** [comment]
@swift-ci please build toolchain

**Comment by hborla** [comment]
@swift-ci please test source compatibility

**Comment by hborla** [comment]
@swift-ci please build toolchain macOS platform

**Comment by hborla** [comment]
@swift-ci please test source compatibility

**Comment by hborla** [comment]
I'm going to split the request refactoring out into a separate PR

**Comment by hborla** [comment]
@swift-ci please build toolchain

**Comment by hborla** [comment]
@swift-ci please test source compatibility

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by tkremenek** [comment]
@swift-ci clean test

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by hborla** [comment]
@swift-ci please test Windows platform

**Pull request merged**